File: scripts/q3.py
import cv2
import numpy as np
import matplotlib.pyplot as plt
import sys

def histograma(img, bins=256):
    hist = np.zeros(bins)

    for px in img.flatten():
        hist[px] += 1
    
    return hist

def cumsum(hist):
    cdf = np.zeros_like(hist)
    cdf[0] = hist[0]

    for i in range(1, len(hist)):
        cdf[i] = cdf[i-1] + hist[i]
    
    return cdf

def equalizacao(img):
    hist = histograma(img)
    cdf = cumsum(hist)

    total_pxs = img.size
    cdf_min = np.min(cdf[cdf>0])

    cdf_norm = (cdf-cdf_min)/(total_pxs-cdf_min) #normalização para [0,1]
    cdf_norm = (cdf_norm*255).astype(np.uint8) #normalização para [0,255]

    img_eq = np.zeros_like(img)

    for x in range(img.shape[0]):
        for y in range(img.shape[1]):
            img_eq[x,y] = cdf_norm[img[x,y]]

    return img_eq, hist, cdf



def main():
    img = cv2.imread(sys.argv[1],cv2.IMREAD_GRAYSCALE)
    img_eq, hist, cdf = equalizacao(img)

    # Histograma original
    plt.subplot(2,3,1)
    plt.bar(range(256), hist)
    plt.title('Histograma Original')

    # Imagem original
    plt.subplot(2,3,2)
    plt.imshow(img, cmap='gray')
    plt.title('Imagem Original')
    plt.axis('Off')

    # Cumsum original
    plt.subplot(2,3,3)
    plt.plot(cdf)
    plt.title('Cumsum Original')

    # Histograma após equalização
    hist_eq = histograma(img_eq)
    plt.subplot(2,3,4)
    plt.bar(range(256), hist_eq)
    plt.title('Histograma após Equalização')

    # Imagem após equalização
    plt.subplot(2,3,5)
    plt.imshow(img_eq, cmap='gray')
    plt.title('Imagem após Equalização')
    plt.axis('off')

    # Cumsum após equalização
    cumsum_eq = cumsum(hist_eq)
    plt.subplot(2,3,6)
    plt.plot(cumsum_eq)
    plt.title('Cumsum após Equalização')


    plt.show()

if __name__ == "__main__":
    main()

File: scripts/q3.cpp
#include <opencv2/opencv.hpp>
#include <iostream>
#include <vector>
#include <algorithm>
#include <cmath>

using namespace cv;
using namespace std;

vector<int> histograma(const Mat& img) {
    vector<int> hist(256, 0);
    for (int r=0; r<img.rows; r++) {
        for (int c=0; c<img.cols; c++) {
            int pixel_val = img.at<uchar>(r, c);
            hist[pixel_val]++;
        }
    }
    return hist;
}

vector<int> cumsum(const vector<int>& hist) {
    vector<int> cdf(256, 0);
    cdf[0] = hist[0];
    for (int i=1; i<256; i++) {
        cdf[i] = cdf[i-1] + hist[i];
    }
    return cdf;
}

void equalizacao(const Mat& img, Mat& img_eq, vector<int>& hist, vector<int>& cdf) {
    hist = histograma(img);
    cdf = cumsum(hist);

    int total_pxs = img.rows * img.cols;
    
    int cdf_min = 0;
    for (int i=0; i<256; i++) {
        if (cdf[i] > 0) {
            cdf_min = cdf[i];
            break;
        }
    }

    vector<uchar> cdf_norm(256, 0);
    for (int i=0; i<256; i++) {
        float valor = (float)(cdf[i] - cdf_min) / (total_pxs - cdf_min); //normalização para [0,1]
        cdf_norm[i] = saturate_cast<uchar>(round(valor * 255.0f)); //normalização para [0,255]
    }

    img_eq = Mat::zeros(img.size(), img.type());
    for (int r=0; r<img.rows; r++) {
        for (int c=0; c<img.cols; c++) {
            img_eq.at<uchar>(r, c) = cdf_norm[img.at<uchar>(r, c)];
        }
    }
}

int main(int argc, char** argv) {

    Mat img = imread(argv[1], IMREAD_GRAYSCALE);

    Mat img_eq;
    vector<int> hist_orig, cdf_orig;
    
    equalizacao(img, img_eq, hist_orig, cdf_orig);

    imshow("Imagem Original", img);
    imshow("Imagem Equalizada", img_eq);
    waitKey(0);

    cv::destroyAllWindows();
    return 0;
}